### PR TITLE
feat(moe): add topk_before_score routing and use_router_bias support

### PR DESCRIPTION
**Pull request by eous**
Add support for GPT-OSS style MoE routing with TopK-then-Score strategy.

Changes to MoEArgs:
- Add topk_before_score: bool = False (preserve backward compat)
- Add use_router_bias: bool = False (for GPT-OSS learned router bias)
- Add use_expert_bias: bool = False (for GPT-OSS expert biases)

Changes to TokenChoiceTopKRouter:
- Support both routing strategies:
  - TopK-then-Score (GPT-OSS): topk on raw logits, then score_func
  - Score-then-TopK (DeepSeek-V3): score_func to all, then topk
- Add validation: node-limited routing not supported with topk_before_score
- Fix expert_bias handling: gather unbiased logits for score computation
- Add _debug_force_load_balance support for TopK-then-Score path
- Add explicit bias initialization in init_weights
- Change histc to bincount for token counting (more reliable)

GPT-OSS config updates:
- 20B/120B: Add topk_before_score=True, use_router_bias=True, use_expert_bias=True
- 120B: Fix top_k from 4 to 8 (correct value for 120B model)
- Add GptOssStateDictAdapter to __all__

### CODE REVIEW

File: torchtitan/models/moe/moe.py
Comment (tianyu-l) [inline torchtitan/models/moe/moe.py:1]: gpt-oss has its own MoE implementation, please don't modify the vanilla one.
https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/gpt_oss/model/moe.py

Comment (eous) [reply torchtitan/models/moe/moe.py:1]: Ok, Ill take a look. Thank you for the pointer.

**Comment by shuhuayu** [comment]
@eous  Thanks for the pr. Can you move the changes to `moe/moe.py` to `gpt_oss/model/moe.py`?

**Comment by shuhuayu** [comment]
@eous I rethink about it. it seems that there is no need to do topk before scores. Now both score functions we support are monotonically increasing functions, so topk before or after does not change the experts we select. For gptoss models, the thing we need to do is set `route_norm = True`, then we do sum normalization over the selected experts softmax scores (over all), this reduces to the same result of softmax over topk experts. However, we do need to add `route_gate_bias` and `expert_bias` options in `MoEArgs` and set them to be `True` for gptoss models. What do you think?

**Comment by eous** [comment]
> @eous I rethink about it. it seems that there is no need to do topk before scores. Now both score functions we support are monotonically increasing functions, so topk before or after does not change the experts we select. For gptoss models, the thing we need to do is set `route_norm = True`, then we do sum normalization over the selected experts softmax scores (over all), this reduces to the same result of softmax over topk experts. However, we do need to add `route_gate_bias` and `expert_bias` options in `MoEArgs` and set them to be `True` for gptoss models. What do you think?

@shuhuayu Thanks, the monotonicity insight simplifies things. I've reworked the PR in [33ec345](https://github.com/pytorch/torchtitan/pull/2212/commits/c9488775574ee5c61fb1af48529b1a06c35e5957) - reverted the vanilla moe.py changes and added a GptOssTokenChoiceTopKRouter subclass that overrides the gate with bias support, plus set route_norm=True as you suggested.                        
                                                                                                                                                    
  One limitation: the subclass has to recreate the gate after the parent already creates it with bias=False, which is a bit awkward and introduces an implicit ordering dependency. Might be cleaner to just plumb use_router_bias into the base TokenChoiceTopKRouter as a backward-compatible param (default False). Let me know if you'd prefer that approach.

---edit
[rank0]:[titan] 2026-01-23 00:33:15,951 - root - INFO - step:  1  loss:  2.2174  grad_norm: 111.5000  memory: 89.57GiB(94.31%)  tps: 292  tflops: 51.40  mfu: 16.47%
[rank0]:[titan] 2026-01-23 00:49:28,102 - root - INFO - step:  5  loss:  1.1087  grad_norm: 20.0000  memory: 89.57GiB(94.31%)  tps: 1,079  tflops: 190.13  mfu: 60.94%
Testing it out on my work branch on my theta train. So far at least just from loss/grad_norm ir seems to be fine.

File: torchtitan/models/gpt_oss/__init__.py [hunk 1]
```diff
@@ -34,13 +35,14 @@
             num_experts=8,
             num_shared_experts=0,
             score_func="softmax",
-            route_norm=False,
+            route_norm=True,  # Normalize over selected experts (match production configs)
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:38]: We can simplify by removing inline comments.

File: torchtitan/models/gpt_oss/__init__.py [hunk 2]
```diff
@@ -34,13 +35,14 @@
...
             route_scale=1.0,
             score_before_experts=False,
             top_k=4,
             use_grouped_mm=True,
             load_balance_coeff=1e-3,
         ),
+        use_router_bias=True,  # GPT-OSS models have learned router gate bias
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:45]: I think it makes better sense to add it as an attribute of the dataclass `MoEArgs`, and name it `use_gate_bias`, also remove the inline comments.

File: torchtitan/models/gpt_oss/__init__.py [hunk 3]
```diff
@@ -49,27 +51,31 @@
             num_experts=32,
             num_shared_experts=0,
             score_func="softmax",
-            route_norm=False,
+            route_norm=True,  # Normalize over selected experts (equivalent to softmax over top-k)
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:54]: ditto.

File: torchtitan/models/gpt_oss/__init__.py [hunk 4]
```diff
@@ -49,27 +51,31 @@
...
             route_scale=1.0,
             score_before_experts=False,
             top_k=4,
             use_grouped_mm=True,
-            load_balance_coeff=1e-3,
+            load_balance_coeff=None,  # Disabled for fine-tuning - not part of official GPT-OSS model
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:59]: revert.

File: torchtitan/models/gpt_oss/__init__.py [hunk 5]
```diff
@@ -49,27 +51,31 @@
...
         ),
+        use_router_bias=True,  # GPT-OSS models have learned router gate bias
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:61]: ditto.

File: torchtitan/models/gpt_oss/__init__.py [hunk 6]
```diff
@@ -49,27 +51,31 @@
...
+        attn_mask_type="causal",
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:62]: revert.

File: torchtitan/models/gpt_oss/__init__.py [hunk 7]
```diff
@@ -49,27 +51,31 @@
...
     ),
     "120b": GptOssModelArgs(
         n_layers=36,
         moe_args=MoEArgs(
             num_experts=128,
             num_shared_experts=0,
             score_func="softmax",
-            route_norm=False,
+            route_norm=True,  # Normalize over selected experts (equivalent to softmax over top-k)
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:70]: ditto.

File: torchtitan/models/gpt_oss/__init__.py [hunk 8]
```diff
@@ -49,27 +51,31 @@
...
             route_scale=1.0,
             score_before_experts=False,
-            top_k=4,
+            top_k=8,  # 120B uses top-8 routing (vs top-4 for 20B)
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:73]: revert. check https://arxiv.org/pdf/2508.10925 section 2.2 2nd paragraph.

File: torchtitan/models/gpt_oss/__init__.py [hunk 9]
```diff
@@ -49,27 +51,31 @@
...
             use_grouped_mm=True,
-            load_balance_coeff=1e-3,
+            load_balance_coeff=None,  # Disabled for fine-tuning - not part of official GPT-OSS model
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:75]: revert.

File: torchtitan/models/gpt_oss/__init__.py [hunk 10]
```diff
@@ -49,27 +51,31 @@
...
         ),
+        use_router_bias=True,  # GPT-OSS models have learned router gate bias
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:77]: ditto.

File: torchtitan/models/gpt_oss/__init__.py [hunk 11]
```diff
@@ -49,27 +51,31 @@
...
+        attn_mask_type="causal",
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/__init__.py:78]: revert.

File: torchtitan/models/gpt_oss/model/args.py [hunk 12]
```diff
@@ -58,6 +58,7 @@ class GptOssModelArgs(BaseModelArgs):
     # MoE
     moe_args: MoEArgs = field(default_factory=MoEArgs)
     swiglu_limit: float = 7.0
+    use_router_bias: bool = False  # GPT-OSS models have learned router gate bias
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/model/args.py:61]: I think it makes better sense to move it inside `MoEArgs`. Renaming it as `use_gate_bias` since we use it in the `self.gate`. And set the default value as `False`. In `gpt_oss/__init__.py`, we set them as `True` within `MoEArgs`.

File: torchtitan/models/gpt_oss/model/moe.py [hunk 13]
```diff
@@ -5,17 +5,65 @@
...
 
 
-from typing import Callable
+from typing import Callable, Literal
 
 import torch
 from torch import nn
 from torch.distributed.tensor import DTensor
-from torchtitan.models.moe.moe import MoE
+from torchtitan.models.moe.moe import MoE, TokenChoiceTopKRouter
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/model/moe.py:13]: revert.

File: torchtitan/models/gpt_oss/model/moe.py [hunk 14]
```diff
@@ -5,17 +5,65 @@
 # LICENSE file in the root directory of this source tree.
 
 
-from typing import Callable
+from typing import Callable, Literal
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/model/moe.py:8]: revert.

File: torchtitan/models/gpt_oss/model/moe.py [hunk 15]
```diff
@@ -5,17 +5,65 @@
...
+        self,
+        dim: int,
+        num_experts: int,
+        num_expert_groups: int | None,
+        num_limited_groups: int | None,
+        top_k: int,
+        score_func: Literal["softmax", "sigmoid"],
+        route_norm: bool,
+        route_scale: float,
+        use_router_bias: bool = True,
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/model/moe.py:41]: I think in this case, to simplify, we can directly add `use_gate_bias` into the `init` in `TokenChoiceTopKRouter` in `moe/moe.py`. In the `MoE` module under `moe/moe.py`, one can pass `moe_args.use_gate_bias` to `self.router` similar to `route_norm`. In this way,  we no longer need to create this new class and share most code.

File: torchtitan/models/gpt_oss/model/moe.py [hunk 16]
```diff
@@ -286,3 +340,19 @@ def __init__(self, model_args: GptOssModelArgs, dim: int, hidden_dim: int):
...
+                num_experts=moe_args.num_experts,
+                num_expert_groups=moe_args.num_expert_groups,
+                num_limited_groups=moe_args.num_limited_groups,
+                top_k=moe_args.top_k,
+                score_func=moe_args.score_func,
+                route_norm=moe_args.route_norm,
+                route_scale=moe_args.route_scale,
+                use_router_bias=True,
+                _debug_force_load_balance=moe_args._debug_force_load_balance,
+            )
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/model/moe.py:358]: revert.

**Comment by shuhuayu** [comment]
@eous Thanks for the rework. Left some comments.

**Comment by eous** [comment]
This looks like its now a duplicate of https://github.com/pytorch/torchtitan/pull/2319, closing it out
